sdk/python/feast/infra/registry/memory.py

```diff
@@ -0,0 +1,758 @@
+import copy
+from typing import Optional, Dict, List, Any, Tuple, Union
+from pathlib import Path
+from datetime import datetime
+
+from feast import usage
+
+# Feast Resources
+from feast.data_source import DataSource
+from feast.entity import Entity
+from feast.feature_service import FeatureService
+from feast.feature_view import FeatureView, BaseFeatureView
+from feast.on_demand_feature_view import OnDemandFeatureView
+from feast.stream_feature_view import StreamFeatureView
+from feast.request_feature_view import RequestFeatureView
+from feast.project_metadata import ProjectMetadata
+from feast.infra.infra_object import Infra
+from feast.saved_dataset import SavedDataset, ValidationReference
+
+from feast.repo_config import RegistryConfig
+
+from feast.protos.feast.core.Registry_pb2 import Registry as RegistryProto
+from feast.infra.registry.base_registry import BaseRegistry
+
+from feast.errors import (
+    ConflictingFeatureViewNames,
+    FeatureServiceNotFoundException,
+    FeatureServiceNameCollisionException,
+    FeatureViewNotFoundException,
+    ValidationReferenceNotFound,
+    RegistryNotBuiltException,
+    EntityNameCollisionException,
+    EntityNotFoundException,
+    DataSourceRepeatNamesException,
+    DataSourceObjectNotFoundException,
+    SavedDatasetNotFound,
+    DuplicateValidationReference,
+    SavedDatasetCollisionException,
+    MissingProjectMetadataException
+)
+
+TimeDependentObject = Union[
+    BaseFeatureView,
+    FeatureView,
+    StreamFeatureView,
+    OnDemandFeatureView,
+    RequestFeatureView,
+    Entity,
+    FeatureService,
+    SavedDataset
+]
+
+FeastResource = Union[
+    TimeDependentObject, DataSource, SavedDataset, ValidationReference, ProjectMetadata, Infra
+]
+
+
+def project_key(project: str, key: str) -> str:
+    # maps (project, key) pair to a single string, called a `projected key`
+    return f"{project}:{key}"
+
+
+def invert_projected_key(projected_key: str) -> Tuple[str, str]:
+    # inverse of `project_key`
+    s = projected_key.split(":")
+    if len(s) != 2:
+        raise ValueError(f"Invalid projected key {projected_key}. Key must follow the format `project:name`.")
+    return s[0], s[1]
+
+
+def list_registry_dict(project: str, registry: Dict[str, FeastResource]) -> List[FeastResource]:
+    # returns all values in registry that belong to `project`
+    return [
+        v for k, v in registry.items() if invert_projected_key(k)[0] == project
+    ]
```

Comment:
Why did you decide to keep the registry flat and have a dict for every FeastResource + projected key instead of having a dict of dicts that we would call like `self.entities["project_1"]["entity_x"]`?

Reply:
Yeah, we can nest this. I don't have a strong preference, I think.